core/src/main/java/org/infinispan/executors/ParallelExecutorHelper.java

```diff
@@ -0,0 +1,59 @@
+package org.infinispan.executors;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+/**
+ * Static class that provides helpers for parallelizing some operations.
+ *
+ * @author wburns
+ * @since 9.0
+ */
+public class ParallelExecutorHelper {
+   private ParallelExecutorHelper() { }
+
+   /**
+    * Useful to allow for a spliterator to be completely decomposed and submitted to the given Executor.  The results
+    * are then fully consumed by the provided Consumer.  The caller can then be notified of completetion by using the
+    * provided CompletableFuture.
+    * @param spliterator the spliterator to decompose and process each entry
+    * @param consumer the consumer that processes each entry
+    * @param executor the executor to submit the tasks to
+    * @param <T> the type of object being processed
+    * @return future that signifies completion of all submitted tasks
+    */
+   public static <T> CompletableFuture<Void> forEach(Spliterator<? extends T> spliterator,
+           Consumer<? super T> consumer, Executor executor) {
+      long size = spliterator.getExactSizeIfKnown();
+      // If the size is unknown we use the builder method to prevent multiple resizings
+      if (size == -1) {
+         // Use stream builder as it requires less resizing than ArrayList if the spliterator is large
+         Stream.Builder<CompletableFuture<Void>> builder = Stream.builder();
+         builder.accept(forEach(spliterator, consumer, executor, builder));
+         return CompletableFuture.allOf((CompletableFuture<Void>[]) builder.build().toArray(i -> new CompletableFuture[i]));
+      } else {
+         if (size > (long) Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("Spliterator has " + size +
+                    " entries, we can't support more than Integer.MAX_VALUE");
+         }
+         // Use stream builder as it requires less resizing than ArrayList if the spliterator is large
+         List<CompletableFuture<Void>> list = new ArrayList<>((int) size);
+         list.add(forEach(spliterator, consumer, executor, list::add));
+         return CompletableFuture.allOf((CompletableFuture<Void>[]) list.toArray(new CompletableFuture[list.size()]));
+      }
+   }
+
+   private static <T> CompletableFuture<Void> forEach(Spliterator<? extends T> spliterator,
+           Consumer<? super T> consumer, Executor executor, Consumer<CompletableFuture<Void>> builder) {
+      Spliterator<? extends T> split;
+      while ((split = spliterator.trySplit()) != null) {
```

Comment:
Shouldn't you stop splitting when the spliterator reaches a certain size to batch consumer calls? Looking at the `ArrayListSpliterator` code, it only stops splitting when it has a single element.